File: src/main/java/org/jcodec/containers/mp4/boxes/CompositionOffsetsBox.java
package org.jcodec.containers.mp4.boxes;

import java.nio.ByteBuffer;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * Default box factory
 * 
 * @author The JCodec project
 * 
 */
public class CompositionOffsetsBox extends FullBox {

    private Entry[] entries;

    public static class Entry {
        public int count;
        public int offset;

        public Entry(int count, int offset) {
            this.count = count;
            this.offset = offset;
        }

        public int getCount() {
            return count;
        }

        public int getOffset() {
            return offset;
        }
    }

    public CompositionOffsetsBox(Header header) {
        super(header);
    }

    public static String fourcc() {
        return "ctts";
    }

    public static CompositionOffsetsBox createCompositionOffsetsBox(Entry[] entries) {
        CompositionOffsetsBox ctts = new CompositionOffsetsBox(new Header(fourcc()));
        ctts.entries = entries;
        return ctts;
    }

    @Override
    public void parse(ByteBuffer input) {
        super.parse(input);
        int num = input.getInt();

        entries = new Entry[num];
        for (int i = 0; i < num; i++) {
            entries[i] = new Entry(input.getInt(), input.getInt());
        }
    }

    @Override
    protected void doWrite(ByteBuffer out) {
        super.doWrite(out);

        out.putInt(entries.length);
        for (int i = 0; i < entries.length; i++) {
            out.putInt(entries[i].count);
            out.putInt(entries[i].offset);
        }
    }

    public Entry[] getEntries() {
        return entries;
    }
}

File: src/main/java/org/jcodec/containers/mp4/boxes/CompositionOffsetsBox.kt
package org.jcodec.containers.mp4.boxes

import java.nio.ByteBuffer

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * Default box factory
 *
 * @author The JCodec project
 */
class CompositionOffsetsBox(header: Header) : FullBox(header) {
    var entries: Array<Entry?> = emptyArray()
        private set

    class Entry(var count: Int, var offset: Int)

    class LongEntry(var count: Long, var offset: Long)

    override fun parse(input: ByteBuffer) {
        super.parse(input)
        val num = input.int
        entries = arrayOfNulls(num)
        for (i in 0 until num) {
            entries[i] = Entry(input.int, input.int)
        }
    }

    override fun doWrite(out: ByteBuffer) {
        super.doWrite(out)
        out.putInt(entries.size)
        for (i in entries.indices) {
            out.putInt(entries[i]!!.count)
            out.putInt(entries[i]!!.offset)
        }
    }

    override fun estimateSize(): Int {
        return 12 + 4 + entries.size * 8
    }

    companion object {
        @JvmStatic
        fun fourcc(): String {
            return "ctts"
        }

        @JvmStatic
        fun createCompositionOffsetsBox(entries: Array<Entry?>): CompositionOffsetsBox {
            val ctts = CompositionOffsetsBox(Header(fourcc()))
            ctts.entries = entries
            return ctts
        }
    }
}